### PR TITLE
Update rollover index API docs for data streams

### PR DESCRIPTION
**Pull request by danhermann**
Documents the behavior of the rollover index API when operating on data streams.

Relates to #53100

### CODE REVIEW

**Review by martijnvg — APPROVED** [review]
LGTM! (pending succesfull build after #55498 gets merged)

**Comment by henningandersen** [comment]
@jrodewig maybe you want to have a look at this too. Given that the terminology (rollover target) will be used in the code too, getting your feedback is valuable.

File: docs/reference/indices/rollover-index.asciidoc [hunk 1]
```diff
@@ -4,8 +4,11 @@
 <titleabbrev>Rollover index</titleabbrev>
 ++++
 
-Assigns an <<indices-aliases, index alias>> to a new index
-when the alias's existing index meets a condition you provide.
+Creates a new index for a rollover target when the target's existing index meets a
+condition you provide. A rollover target can be either an <<indices-aliases, index alias>>
+or a <<indices-datastreams, data stream>>. When targeting an alias, the alias is updated to
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:9]: `indices-datastreams` isn't a valid anchor in the docs right now. This is causing the docs CI to fail.

> or a <<indices-datastreams, data stream>>. When targeting an alias, the alias is updated to

It looks like we don't have data stream docs at all atm. It might be good to add those before merging this.

File: docs/reference/indices/rollover-index.asciidoc [hunk 2]
```diff
@@ -4,8 +4,11 @@
 <titleabbrev>Rollover index</titleabbrev>
 ++++
 
-Assigns an <<indices-aliases, index alias>> to a new index
-when the alias's existing index meets a condition you provide.
+Creates a new index for a rollover target when the target's existing index meets a
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:7]: 👍  I feel like 'rollover target' is a good term for this and it's precisely explained here.

File: docs/reference/indices/rollover-index.asciidoc [hunk 3]
```diff
@@ -4,8 +4,11 @@
...
+condition you provide. A rollover target can be either an <<indices-aliases, index alias>>
+or a <<indices-datastreams, data stream>>. When targeting an alias, the alias is updated to
+reference the new index. When targeting a data stream, the new index becomes the data stream's
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:10]: nit: I think "point" is a bit more clear here.
```suggestion
point to the new index. When targeting a data stream, the new index becomes the data stream's
```

File: docs/reference/indices/rollover-index.asciidoc [hunk 4]
```diff
@@ -25,15 +28,15 @@ POST /alias1/_rollover/twitter
...
+`POST /<rollover-target>/_rollover/`
 
 
 [[rollover-index-api-desc]]
 ==== {api-description-title}
 
-The rollover index API rolls an <<indices-aliases, alias>> to a new index when
+The rollover index API rolls an existing index for a rollover target to a new index when
 the existing index meets a condition you provide. You can use this API to retire
 an index that becomes too large or too old.
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:41]:
```suggestion
The rollover index API rolls a rollover target's existing index to a new index
when the existing index meets a condition you provide. You can use this API to 
retire an index that becomes too large or too old.
```

File: docs/reference/indices/rollover-index.asciidoc [hunk 5]
```diff
@@ -73,10 +81,10 @@ index creation applies to the rollover action.
 [[rollover-index-api-path-params]]
 ==== {api-path-parms-title}
 
-`<alias>`::
+`<rollover-target>`::
 (Required, string)
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:85]: For parameters that are required in some situations, we typically add an asterisk.
```suggestion
(Required*, string)
```

File: docs/reference/indices/rollover-index.asciidoc [hunk 6]
```diff
@@ -127,7 +132,7 @@ include::{docdir}/rest-api/common-parms.asciidoc[tag=aliases]
 +
 --
 (Required, object)
-Set of conditions the index alias's existing index must met to roll over.
+Set of conditions the rollover target's existing index must meet to roll over.
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:135]: Good catch!

File: docs/reference/indices/rollover-index.asciidoc [hunk 7]
```diff
@@ -206,13 +211,69 @@ The API returns the following response:
 <2> Whether the rollover was dry run.
 <3> The result of each condition.
 
+
+[[rollover-data-stream-ex]]
+===== Roll over a data stream
+
+[source,console]
+--------------------------------------------------
+PUT _data_stream/my-data-stream <1>
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:220]:
```suggestion
PUT /_data_stream/my-data-stream <1>
```

File: docs/reference/indices/rollover-index.asciidoc [hunk 8]
```diff
@@ -42,25 +45,30 @@ NOTE: To roll over an index, a condition must be met *when you call the API*.
...
+In this case, the rollover request:
 
 . Creates a new index
 . Sets `is_write_index` to `true` for the new index
 . Sets `is_write_index` to `false` for the original index
 
+If the specified rollover target is a data stream, the rollover request:
+
+. Creates a new index
+. Adds the new index as a backing index on the data stream
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:69]: Should we mention that it's also the write index?

File: docs/reference/indices/rollover-index.asciidoc [hunk 9]
```diff
@@ -206,13 +211,69 @@ The API returns the following response:
...
+  "max_size": "5gb"
+  }
+}
+--------------------------------------------------
+<1> Creates a data stream called `my-data-stream` with one initial backing index
+named `my-data-stream-000001`.
+<2> If the current index of `my-data-stream` was created 7 or more days ago, or
+contains 1,000 or more documents, or has an index size at least around 5GB, then
+a new backing index `my-data-stream-000002` index is created and added to the
+`my-data-stream` data stream.
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:241]: The conditions become a bit hard to parse in a callout. I think breaking them out into a bulleted list might help>
```suggestion
<2> This request creates a new backing index, `my-data-stream-000002`, and adds
it as the write index for the `my-data-stream` data stream if the current
write index meets at least one of the following conditions:
+
--
* The index was created 7 or more days ago.
* The index has an index size of 5GB or greater.
* The index contains 1,000 or more documents.
--
```

File: docs/reference/indices/rollover-index.asciidoc [hunk 10]
```diff
@@ -241,11 +302,14 @@ POST /logs_write/_rollover
 
 [[rollover-index-specify-index-ex]]
 ===== Specify a target index name
+If the rollover target is a data stream, you may not specify a target index
```
Comment (jrodewig) [inline docs/reference/indices/rollover-index.asciidoc:305]: nit
```suggestion
If the rollover target is a data stream, you cannot specify a target index
```

**Review by jrodewig — APPROVED** [review]
Look good overall. I mostly left minor wording nits, but feel free to ignore those as you see fit.

I think the rollover target terminology works well, and you did a good job of explaining it concisely upfront.

Here are some other things I wanted to point out:

* We don't currently have any data stream docs. For the best user experience, it may make sense to wait until those are added before merging this PR.
* The `indices-datastreams` anchor linked here doesn't exist. This is likely causing the docs CI build to fail and is related to by first point.
* At several points, the line length exceeds 80 characters. Please rewrap those where possible.

File: docs/reference/indices/rollover-index.asciidoc (hunk 1 above)
Comment (danhermann) [reply docs/reference/indices/rollover-index.asciidoc:9]: @jrodewig, I've added those in https://github.com/elastic/elasticsearch/pull/55557 if you would like to see them. I presume they will need to be merged first to create the appropriate data streams anchors.

**Comment by danhermann** [comment]
@jrodewig, thanks for the detailed review (as well as corrected the optionality of the `conditions` option which was incorrect). I've made your suggested changes. I think I'll need to get https://github.com/elastic/elasticsearch/pull/55551 merged first so the anchors will resolve.

File: docs/reference/indices/rollover-index.asciidoc (hunk 1 above)
Comment (jrodewig) [reply docs/reference/indices/rollover-index.asciidoc]: 👍 Awesome. And yes--- they'll need to be merged in before you can reference the anchor here. You may have to update this branch after #55557 merges to get the docs CI to pass.

**Comment by danhermann** [comment]
@elasticmachine update branch
